src/librustc/dep_graph/graph.rs

```diff
@@ -676,18 +654,23 @@ impl DepGraph {
                             continue;
                         }
                     } else {
+                        // FIXME: This match is just a workaround for incremental bugs and should be removed.
                         match dep_dep_node.kind {
-                            DepKind::Hir | DepKind::HirBody | DepKind::CrateMetadata => {
+                            DepKind::hir_owner
+                            | DepKind::hir_owner_items
+                            | DepKind::CrateMetadata => {
```

Comment:
I see how this is still relevant for `DepKind::CrateMetadata` but I'm surprised that it's needed for `DepKind::hir_owner` and `DepKind::hir_owner_items`.

Reply:
Given that `hir_owner` is just a rename of `Hir` and `hir_owner_items` is just a rename of `HirBody`, you can follow the [explanation here](https://github.com/rust-lang/rust/issues/62649#issuecomment-589018098) to see how this has the same problem as `Hir` and `HirBody`. Also `Hir(Something)` will be replaced by `hir_owner_items(Something)` there, but that doesn't affect anything.

Reply:
I had a chance to test this and I think the explanation in https://github.com/rust-lang/rust/issues/62649#issuecomment-589082343 basically applies here:

- `Something::foo` is a valid query key for `hir_owner` in `rpass1` but in `rpass2` it's not a valid key anymore. 
- The red-green algorithm does not detect this because `RecoverKey::recover()` returns `Some(_)` instead of `None`. The algorithm thus goes ahead with forcing `DepNode::hir_owner(Something::foo)` and ICEs.

A short-term fix for this would be to have a newtype `struct HirOwnerDefId(DefId)` that is a `DefId` with the additional invariant that it must correspond to a hir-owner. Then `<HirOwnerDefId as RecoverKey>::recover()` could apply the same check as done [here](https://github.com/rust-lang/rust/blob/2cb0b8582ebbf9784db9cec06fff517badbf4553/src/librustc/dep_graph/graph.rs#L918-L921) and return `None` if it doesn't hold. Most uses of `HirOwnerDefId` should stay hidden behind the hir map wrapper, so this shouldn't be too disruptive.

Another workaround would be to diversify the `DefPathData` variants again so that a field and a method would not have the same `DefPath` anymore. Either fixes would be acceptable solutions for unblocking this PR, in my opinion. 

I think one way to fix the problem at the conceptual level would be to add the invariant that `RecoverKey::recover()` must only return `Some(_)` if the recovered key is actually a valid query-key for the given query. That is, the result of `RecoverKey::recover()` would need to depend on the query, not just on the type of the query-key. It would then be able to do query-specific runtime checks about key validity.

Reply:
This doesn't block this PR though. It's just https://github.com/rust-lang/rust/pull/68289 ported to it. There's no further problems caused by this PR compared to master.

It seems like you're trying to fix https://github.com/rust-lang/rust/issues/62649 here. I mentioned the `DefPathData` solution in a [comment](https://github.com/rust-lang/rust/issues/62649#issuecomment-585355880)) there. I want to check the performance of making `as_local_hir_id` a query, but that is blocked on this PR.

Reply:
Well, I'm fine with merging this if you mention #62649 in the FIXME comment above.

Reply:
I've added a reference to that issue.